pkg/logs/input/file/tailer_key.go

```diff
@@ -0,0 +1,19 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+package file
+
+import "fmt"
+
+// tailerKey can be implemented by File and Tailer
+type tailerKey interface {
```

Comment:
I agree that this would be the way to go (Tailer should contain File, they have a hard dependency in this case and that's way easier to think this way).

I've done it in a branch and overall I feel like the result is fine, slightly more clear but I'm afraid it's too big of a change if we want to ship this tomorrow, see the commit here: https://github.com/DataDog/datadog-agent/commit/59bd1969fd12c98bd13e1c29e2ebc698dd353feb It introduces the concept the the Scanner needs a key, which is way more clear to me than mentioning some source configs.

This is one in many refactor that we should consider within these sources, however being this large, I would vote for pushing this into master later-ish.